lib/stdlib/src/gen_server.erl

```diff
@@ -692,6 +672,62 @@ See [`erlang:start_timer/4`](`erlang:start_timer/4`) for details.
 """.
 -type timeout_option() :: {abs, Abs :: boolean()}.
 
+-doc """
+Callback and server loop action.
+
+Returned by the callbacks [`Module:init/1`](`c:init/1`),
+[`Module:handle_call/3`](`c:handle_call/3`),
+[`Module:handle_cast/2`](`c:handle_cast/2`),
+and [`Module:handle_continue/2`](`c:handle_continue/2`).
+It is also the last argument to [`enter_loop/4,5`](`enter_loop/4`).
+
+`t:action/0` is one of:
+
+- **`Time :: `[`timeout()`](`t:timeout/0`)** -
+  If the action is an integer `Time`, a time-out occurs
+  unless a request or a message is received within that many milliseconds.
+  A time-out is represented by the atom `timeout` as the `Info` argument
+  to be handled by the [`Module:handle_info/2`](`c:handle_info/2`)
+  callback function.  The action `infinity` can be used to wait indefinitely,
+  which is the default when there is no `t:action/0` specified.
+
+  A system message restarts the time-out, which is a known
+  and unfortunate flaw in its implementation.  This also applies to
+  stray (cancelled) timer messages from the
+  `{timeout|hibernate, ...}` time-outs described below,
+  so it is recommended to not use them
+  in combination with this legacy time-out type.
+
+- **`{timeout, Time, Message}`\
+  `{timeout, Time, Message, Options}`** - Like `Time` above,
+  but the delivered `Info` argument is specified by `Message`,
+  and the time-out is not affected by system messages.
+
+  `Options` can be used to trigger the time-out at an absolute point in time
+  instead.  See `t:timeout_option/0` and `erlang:start_timer/4` for details.
+
+  A relative time-out with `Time = 0` is immediately delivered,
+  before any request or message is received, including system messages.
+
+  A time-out with `Time = infinity` will never be delivered
+  so the action is ignored; no time-out is started.
+
+- **`{hibernate, Time, Message}`\
+  `{hibernate, Time, Message, Options}`** - A combination of
+  the action `hibernate` below, and  `{timeout, Time, Message}`
+  or `{timeout, Time, Message, Options}` above.
+
+  The process goes into hibernation while waiting for the next
+  request or message to arrive, or for the time-out to expire.
+
+- **`hibernate`** - The process goes into hibernation
+  (by calling `erlang:hibernate/0`), waiting for the next
+  request or message to arrive
```

Comment:
The `gen_statem` docs mentions the `garbage_collect` optimization. Maybe do that here, too? It is not strictly necessary to mention, though :man_shrugging: